C++/Navio/Common/gpio.cpp

```diff
@@ -137,37 +132,48 @@ void Pin::toggle()
     write(!read());
 }
 
+uint32_t Pin::getPeriBase() const
+{
+    int rpi_version = getRaspberryPiVersion();
+
+    switch (rpi_version) {
+      case 1:
+        return BCM2708_PERI_BASE;
+      case 2:
+        return BCM2709_PERI_BASE;
+      case 3:
+        return BCM2835_PERI_BASE;
+      default:
+        fprintf(stderr, "Unexpected version. Setting up configurations as on RPI2/3\n");
+        return BCM2709_PERI_BASE;
+}
+}
+
+
+
 int Pin::getRaspberryPiVersion() const
 {
-    char buffer[MAX_SIZE_LINE];
-    const char* hardware_description_entry = "Hardware";
-    const char* v1 = "BCM2708";
-    const char* v2 = "BCM2709";
-    const char* v3 = "BCM2835";
-    char* flag;
-    FILE* fd;
-
-    fd = fopen("/proc/cpuinfo", "r");
-
-    while (fgets(buffer, MAX_SIZE_LINE, fd) != NULL) {
-        flag = strstr(buffer, hardware_description_entry);
-
-        if (flag != NULL) {
-            if (strstr(buffer, v3) != NULL) {
-                fclose(fd);
-                return 3;
-            } else if (strstr(buffer, v2) != NULL) {
-                fclose(fd);
-                return 2;
-            } else if (strstr(buffer, v1) != NULL) {
-                fclose(fd);
-                return 1;
-            }
+    int rpi_version;
+
+    size_t i = 0;
+    std::string buffer;
+    std::ifstream f(SYSFS_RPI_MODEL_FILE_PATH);
+
+    if (f.is_open()){
+      while (std::getline(f, buffer)) {
+        while (!isdigit(buffer[i])) { //first digit is a number of RPI version (from 1 to 4)
+          i++; //skip
         }
-    }
+        char version_char = buffer[i];
+        rpi_version = atoi(&version_char);
+
+        return rpi_version;
+     }
+        fprintf(stderr, "Could not read RPi model, defaulting to 1\n");
+        return 1;
+     }
+
+     fprintf(stderr, "Could not open ", SYSFS_RPI_MODEL_FILE_PATH, "defaulting to 1\n");
+     return 1;
```

Comment:
1. Invert the logic (random [link](https://www.geeksforgeeks.org/writing-clean-else-statements/) from Google).
2. "defaulting to" - bad decision. Just drop it.

Reply:
3. return 1 - bad decision. Add a `#define` for exit status.

Reply:
4. Why fprintf? Am i wrong or `std::cerr <<` is how we do it in CPP?

Reply:
2. Shouldn't I inform users which model I will use if there would be an error?
3. Is `DEFAULT_RPI_VERSION` good for this define?
4. Yes, it just was in original code and I followed it.

Reply:
If there would be an error you do nothing but tell the user about it and exit with an error.

Reply:
Get it. I thought we don't exit with it, just setting something else.

Reply:
1. Look for it, please. I change the logic in the first case (f.is_open) but left the second one as it is because there's one common error. Should I invert it too?